README.md
<div align="center">

# ![TERMINALLI](./media/logo.svg "Terminalli")

[![[ INSTALL ]](https://img.shields.io/badge/INSTALL-npm%20i%20termialli-ff0000.svg?style=flat-square&logo=npm)](https://www.npmjs.com/package/terminalli "View NPM package")
[![ LICENSE ](https://img.shields.io/github/license/Benjalaazshah/Terminalli?color=008000&label=LICENSE&style=flat-square)](./LICENSE "View license")
[![ VERSION ](https://img.shields.io/npm/v/terminalli?color=ff8000&label=VERSION&style=flat-square)](./CHANGELOG.md "View changelog")

Still using `console.log` in 2022?! With Terminalli, you can attach timestamps to your logs, organize your logs into different categories (levels) with their own unique color and more!

</div>

![TERMINALLI PREVIEW IMAGE](./media/preview_image.svg "Preview Image")

## FEATURES

### **Features include:**

-   Ability to have each log show the exact date and time when it was sent to the console.
-   Ability to have _relative_ timestamps in your logs. These include timestamps relative to when the previous message was logged to the console.
-   Being able to add or remove any of these types of timestamps from your logs.
-   ![[ NEW ]](https://img.shields.io/badge/-NEW-008000.svg?style=square "New") You are now able to log to up to 6 different preset levels/categories, all of which have there own color.
-   ![[ NEW ]](https://img.shields.io/badge/-NEW-008000.svg?style=square "New") As well as that, you can create custom levels with whatever color you desire!

## BASIC USAGE

First, import the `Terminal` class from Terminalli.

```js
import Terminal from "terminalli";
```

Next up is to create your terminal.

```js
const terminal = new Terminal({
	levels: [
		// Levels...
	]
});
```

Right inside of `levels` is where you are free to define your different levels/categories that you can log to. But to save time, you can import a bunch of preset levels like this:

```js
import Terminal, { advancedLevels } from "terminalli";

const terminal = new Terminal({ levels: advancedLevels });
```

> ![[ NOTE ]](https://img.shields.io/badge/-NOTE-000080.svg?style=square "Note") If you want to learn how to create custom levels, continue reading [here](./README.md#custom-levels 'View "CUSTOM LEVELS" section').

To start logging things, do so like this:

```js
terminal.log.debug("Initial log.");
```

`debug` in this case is the name of one of the preset levels that you can log to. If you were logging an error, for example, you can use the `error` level.

```js
terminal.log.error("Something happened.");
```

### **Output:**

![OUTPUT IMAGE](./media/basic_usage_output_image.svg "Output")

## CUSTOMIZABILITY

Upon creating a new instance of the `Terminal` class, you can turn on and off a bunch of options that determine how logs are to be displayed in the terminal. In the following example, there are three instances of the `Terminal` class all with different settings.

### **Code:**

```js
import Terminal, { advancedLevels } from "terminalli";

const t1 = new Terminal({
	levels: advancedLevels
});

const t2 = new Terminal({
	levels: advancedLevels,
	showDate: true,
	showLevelName: true,
	showTimestampRelativeToLastLog: false,
	use24HourClock: true
});

const t3 = new Terminal({
	levels: advancedLevels,
	showArrow: true,
	showRelativeTimestamp: true,
	showTimestamp: false
});

t1.log.warn("The fox is coming...");
t2.log.warn("The fox is coming...");
t3.log.warn("The fox is coming...");
```

### **Output:**

![OUTPUT IMAGE](./media/customizability_output_image.svg "Output")

## CUSTOM LEVELS

Creating custom levels is as simple as it gets! In the following example, there are two additional custom levels, alongside the preset levels, called `weather` and `victory`.

### **Code:**

```js
import Terminal, { advancedLevels } from "terminalli";

const terminal = new Terminal({
	showLevelName: true,

	levels: [
		...advancedLevels,
		{
			color: ["bgGreen", "white", "bold"],
			isError: false,
			name: "victory"
		},
		{
			color: ["cyan", "bold", "underline"],
			isError: false,
			name: "weather"
		}
	]
});

terminal.log.weather("Today will be 280°C (about the same as an oven).");
terminal.log.victory("The octopus has won.");
```

### **Output:**

![OUTPUT IMAGE](./media/custom_levels_output_image.svg "Output")

## CONTRIBUTING

Found a big fat bug in the code? Have a feature suggestion? [Open an issue here](https://github.com/Benjalaazshah/Terminalli/issues "View all issues on GitHub")! Pull requests are also welcome. Create a pull request [here](https://github.com/Benjalaazshah/Terminalli/pulls "View all pull requests on GitHub")!

---

<div align="center">

## LICENSE

<h6>

<sup><sub>MIT License</sup></sub>

<sup><sub>Copyright (c) 2022 Benjalaazshah</sup></sub>

<sup><sub>Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:</sup></sub>

<sup><sub>The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.</sup></sub>

<sup><sub>THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.</sup></sub>

</h6>
</div>


CHANGELOG.md
# WHAT'S NEW IN v2.0.0

## NEW MAJOR FEATURES

-   You now have the ability to log to up to 6 different preset levels.
-   Creating custom levels is now possible.
-   You can now leave the level name attached to logs to be lowercase.
-   You are now able to add an arrow between the timestamp of your logs and the message of your logs.

## NEW MINOR FEATURES

-   Added new type export `Color`.
-   Added new type export `Level`.
-   Added new type export `Locked`.

## CHANGES

-   Changed defaults for parameter `data` of the `Terminal` class.


sources/types.ts
import { Kleur } from "kleur";

/**
 * Represents an ANSI color.
 */
export type Color = keyof Kleur;

/**
 * Make every property and sub-property read-only.
 */
export type Locked<T> = { readonly [K in keyof T]: T[K] extends object ? Locked<T[K]> : T[K] };

/**
 * Same as `TerminalConstructorData<string>`, but all properties are required and read-only. This interface is used for the `data` property in the `Terminal` class.
 */
export type TerminalData = Locked<Required<TerminalConstructorData<string>>>;

/**
 * Represents a category to which you can log. Levels are usually used as categories of importance.
 */
export interface Level<L extends string> {
	/***
	 * Any ANSI colors/formats which you want any messages logged to this level to have their timestamps highlighted in.
	 */
	color: Color[];

	/**
	 * Whether or not should each message logged to this level be sent to `process.stderr` (standard error stream) instead of `process.stdout` (standard out stream).
	 */
	isError: boolean;

	/**
	 * The name of this level.
	 */
	name: L;
}

/**
 * Customization options for the `Terminal` class.
 */
export interface TerminalConstructorData<L extends string> {
	/**
	 * Whether or not to capitalize the name of a log's corresponding level when it's attached to the log. Keep in mind that you can only see the name of a log's corresponding level when `showLevelName` is also true.
	 *
	 * **No capitalization:**
	 *
	 * `[ info ] Something very interesting happened.`
	 *
	 * **Capitalization:**
	 *
	 * `[ INFO ] Something very interesting happened.`
	 */
	capitalizeLevelName?: boolean;

	/**
	 * A list of categories which you can log to.
	 */
	levels: Level<L>[];

	/**
	 * Whether or not to show a cool arrow before a log's message.
	 *
	 * `>> baz`
	 */
	showArrow?: boolean;

	/**
	 * If true, each message logged to the terminal will have a date corresponding to when the message was logged attached to it.
	 *
	 * `[ 12d/5m/2011y ] foo`
	 */
	showDate?: boolean;

	/**
	 * If true, each message logged to the terminal will have the name of the level of the message attached to it.
	 *
	 * `[ fatal ] WHAT WILL I DO?!`
	 */
	showLevelName?: boolean;

	/**
	 * If true, the date on each message logged to the console (which only exists if `showDate` is also true) will have the month before the day.
	 *
	 * **Day before month:**
	 *
	 * `[ 20d/12m/1999y ] loy`
	 *
	 * **Month before day:**
	 *
	 * `[ 12m/20d/1999y ] loy`
	 */
	showMonthBeforeDay?: boolean;

	/**
	 * If true, each message logged to the terminal will have a timestamp relative to the creation of this particular instance of the `Terminal` class to which this message has been logged.
	 *
	 * `[ 5y 1m 15h 51min 7s 300ms ] A long time has passed.`
	 */
	showRelativeTimestamp?: boolean;

	/**
	 * If true, each message logged to the terminal will have a timestamp corresponding to the exact time the message was logged.
	 *
	 * `[ 13:43:10.23 ] bar`
	 */
	showTimestamp?: boolean;

	/**
	 * If true, each message logged to the terminal will have a timestamp relative to when the previous message was logged to the terminal.
	 *
	 * `[ +31min +5s +903ms ] It took forever!`
	 */
	showTimestampRelativeToLastLog?: boolean;

	/**
	 * If true, the timestamp on each message logged to the console (which only exists if `showTimestamp` is true) will be displayed using the 24 hour clock instead of the 12 hour clock.
	 *
	 * **24 hour clock:**
	 *
	 * `[ 13:27:55.33 ] pow`
	 *
	 * **12 hour clock:**
	 *
	 * `[ 1:27:55.33 PM ] pow`
	 */
	use24HourClock?: boolean;
}


sources/index.ts
import C from "kleur";
import { Color, Level, Locked, TerminalConstructorData, TerminalData } from "./types";

export { Level, Locked, TerminalConstructorData, TerminalData };

/**
 * A couple of preset levels. This is useful for a basic application.
 *
 * **Levels include:**
 *
 * -   error
 * -   trace
 */
export const basicLevels: Level<"error" | "trace">[] = [
	{ color: ["red", "underline"], name: "error", isError: true },
	{ color: ["black", "bold"], name: "trace", isError: false }
];

/**
 * A list of preset levels that you can use to log messages of various levels of importance.
 *
 * **Levels include:**
 *
 * -   debug
 * -   error
 * -   fatal
 * -   info
 * -   trace
 * -   warn
 */
export const advancedLevels: Level<"debug" | "error" | "fatal" | "info" | "trace" | "warn">[] = [
	...basicLevels,
	{ color: ["black", "bold", "underline"], isError: false, name: "debug" },
	{ color: ["bgRed", "white", "bold", "italic"], isError: true, name: "fatal" },
	{ color: ["blue", "bold"], isError: false, name: "info" },
	{ color: ["yellow", "dim", "underline"], isError: true, name: "warn" }
];

/**
 * Represents the console.
 */
export default class Terminal<L extends string> {
	/**
	 * Customization options that were inputted when this terminal instance was created.
	 */
	readonly data: TerminalData;

	/**
	 * Represents the logger. Any methods of `log` logs a message to a specific level.
	 *
	 * @example
	 *
	 * ```
	 * log.error("faz");
	 * ```
	 *
	 * Logs "faz" to the `error` level if such a level even exists.
	 */
	readonly log: Record<L, (message: string) => void>;

	/**
	 * The time when this terminal instance was created.
	 */
	readonly startTime: Date;

	/**
	 * The time when the last message was logged to the terminal.
	 */
	timeInLastLog: Date;

	private _log(level: Level<string>, message: string) {
		const {
			capitalizeLevelName,
			showArrow,
			showDate,
			showLevelName,
			showMonthBeforeDay,
			showRelativeTimestamp,
			showTimestamp,
			showTimestampRelativeToLastLog,
			use24HourClock
		} = this.data;

		const time = new Date();
		const color = getColorApplier("TEXT", "white");
		const colorBG = getColorApplier("BACKGROUND", "bgBlack");
		let monthPositionSwitch = showMonthBeforeDay;
		let output = "";

		function formatChangeInTime(from: Date, prefix: string) {
			let formattedChangeInTime = " ";
			let remainder = time.getTime() - from.getTime();

			function addUnitOfTime(unitValueInMilliseconds: number, unitName: string) {
				const unitCount = Math.floor(remainder / unitValueInMilliseconds);

				remainder = remainder % unitValueInMilliseconds;
				formattedChangeInTime += unitCount !== 0 ? color(prefix + unitCount + unitName) + " " : unitValueInMilliseconds === 1 ? color(prefix + "0") + " " : "";
			}

			addUnitOfTime(31536000000, "y"); // YEARS
			addUnitOfTime(2592000000, "m"); // MONTHS
			addUnitOfTime(86400000, "d"); // DAYS
			addUnitOfTime(3600000, "h"); // HOURS
			addUnitOfTime(60000, "min"); // MINUTES
			addUnitOfTime(1000, "s"); // SECONDS
			addUnitOfTime(1, "ms"); // MILLISECONDS

			return colorBG(formattedChangeInTime);
		}

		function formatMonth() {
			monthPositionSwitch = !monthPositionSwitch;
			return monthPositionSwitch ? "" : `${time.getMonth()}m/`;
		}

		function getColorApplier(colorType: "BACKGROUND" | "TEXT", altColor: Color) {
			const colors = level.color.filter((colorString) => {
				const isBackground = colorString.startsWith("bg");

				return colorType === "BACKGROUND" ? isBackground : !isBackground;
			});

			let color = C[colors[0] || altColor];

			for (let i = 1; i < colors.length; i++) color = color()[colors[i]];
			return color;
		}

		// Should look like: [ ERROR ] or [ error ]
		if (showLevelName) output += `[${colorBG(" " + color(capitalizeLevelName ? level.name.toUpperCase() : level.name) + " ")}]\t`;

		// Should look like: [ 12d/5m/2011y | 13:43:10.23 ] or [ 5m/12d/2011y | 1:43:10.23 PM ]
		if (showDate || showTimestamp) {
			output += "[";
			if (showDate) output += colorBG(" " + color(`${formatMonth() + time.getDate()}d/${formatMonth()}${time.getFullYear()}y`) + " ");
			if (showDate && showTimestamp) output += "|";

			if (showTimestamp) {
				const hours = time.getHours();

				output += colorBG(
					" " +
						color(
							`${use24HourClock || !(hours >= 13 || hours === 0) ? hours : Math.abs(hours - 12)}:${time.getMinutes()}:${time.getSeconds()}.${time.getMilliseconds()}`
						) +
						" " +
						(use24HourClock ? "" : color(hours >= 13 ? "PM" : "AM") + " ")
				);
			}

			output += "]";
		}

		if ((showDate || showTimestamp) && (showRelativeTimestamp || showTimestampRelativeToLastLog)) output += "\t";

		// Should look like: [ 5y 1m 15h 51min 7s 300ms | +31min +5s +903ms ]
		if (showRelativeTimestamp || showTimestampRelativeToLastLog) {
			output += "[";
			if (showRelativeTimestamp) output += formatChangeInTime(this.startTime, "");
			if (showRelativeTimestamp && showTimestampRelativeToLastLog) output += "|";
			if (showTimestampRelativeToLastLog) output += formatChangeInTime(this.timeInLastLog, "+");
			output += "]";
		}

		// Should look like: >>
		if (showArrow) output += ` ${C.bold(">>")}\t`;

		output += `\t${message}\n`;
		(level.isError ? process.stderr : process.stdout).write(output);
		this.timeInLastLog = time;
	}

	/**
	 * Represents the console.
	 *
	 * @param data Any customization options for the terminal.
	 */
	constructor(data: TerminalConstructorData<L>) {
		const defaultData: Omit<TerminalData, "levels"> = {
			capitalizeLevelName: true,
			showArrow: false,
			showDate: false,
			showLevelName: false,
			showMonthBeforeDay: false,
			showRelativeTimestamp: false,
			showTimestamp: true,
			showTimestampRelativeToLastLog: true,
			use24HourClock: false
		};

		let logger: object = {};
		let registeredLevels: string[] = [];

		this.startTime = new Date();
		this.timeInLastLog = this.startTime;
		this.data = Object.assign({}, defaultData, data);

		for (const level of data.levels) {
			if (registeredLevels.some((value) => value === level.name)) throw new Error(`Duplicate level name "${level.name}".`);
			registeredLevels = [...registeredLevels, level.name];

			logger = {
				...logger,

				[level.name]: (message: string) => {
					this._log(level, message);
				}
			};
		}

		this.log = logger as Record<L, (message: string) => void>;
	}
}
